### DISCUSSION TITLE
Lf -single and custom copy/move

### DISCUSSION THREAD

**Discussion by og900aero** [gokcehan/lf#1618] [Q&A] [answered]
I use custom copy/move version:

```
cmd toggle_new &{{
	if [ $(cat ~/.local/share/lf/select_files | grep -c $f) -gt 0 ]; then
		sed -i "\\:$f:d" ~/.local/share/lf/select_files
	else
		echo $f >> ~/.local/share/lf/select_files
	fi
}}

cmd reset_select_files &{{
	truncate -s 0 ~/.local/share/lf/select_files
}}

cmd signalMoveOrCopy &{{
	lf -remote "send unselect"
	cat ~/.local/share/lf/select_files | while read -r line
	do
        lf -remote "send toggle \"$line\""
    done
	lf -remote "send $id $1"
	if sudo -u "$USER" test -w "$PWD"; then
		lf -remote "send $id paste"
	else
		sudo -A lf -remote "send $id paste"
	fi
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reset_select_files"
}}

map <space> :toggle_new; toggle; down
map p :signalMoveOrCopy copy
map P :signalMoveOrCopy cut
```

But doesnt work. The selected files/directories writes to the select_files file, but the copy and move command not execute.
When run lf without -single switch, then work fine, but with -single, not working anymore.
What is the problem?

**Comment by gokcehan** [comment, marked as answer]
@og900aero Remote commands are not supported with `-single`.

**Reply by og900aero** [reply]
Then I think there's not any way around this, I have to use it in server-client mode?

**Reply by gokcehan** [reply]
@og900aero I haven't looked at your examples in detail. If it is possible to achive what you want without remote commands, then you can use `-single`, otherwise I don't see any other way.

**Reply by joelim-work** [reply]
Regular commands (i.e. `cmd foo :{{ ... }}`) are good for chaining multiple commands in sequence, but otherwise the syntax is rather limited. More advanced requirements like control flow (`if` statements and loops), and evaluating environment variables have to be done as a shell command, using `lf -remote` to call back into `lf`, but then this requires the use of the server.

**Reply by og900aero** [reply]
joelim-work thx for your help. I begin reworked my lfrc code, but...with such limitations, it's quite difficult to rewrite functions I've already done. If I only want to run one copy (-single), then it is really quite limited. I'm so sorry for this :(

**Reply by og900aero** [reply]
Can this be solved in this spirit? Because I should write $, &, % in front of a shell command, but here there is an if command, which is completely confusing. Or can it not be solved like this anymore, because the limitation has an impact on this as well?

```
cmd toggle_preview :{{
	if [ "$lf_preview" = "true" ]; then 
		set preview false
		set ratios 1
	else 
		set ratios 1:1
		set preview true
	fi
}}
```

**Reply by gokcehan** [reply]
@og900aero May I ask, what is your rationale for the use of `-single`? It is only meant for people using little or no configuration and/or only a single client. It seems that you want to configure the program, so why do you want to use `-single`? In your case, it seems to work as an artificial limitation that you put on yourself.

**Reply by og900aero** [reply]
@gokcehan The reason I would use it is that I don't want to run multiple instances. 1 is enough. I solved it with modifications so that I don't need tabs or a server-client version. The server-client version is not suitable for me anyway, because it takes up unnecessary extra space on the screen (it also runs more than I need in the process). However, the fact that I only want to run 1 lf doesn't rule out the fact that I don't want to customize it as much as possible. This is exactly the essence of lf, that it can be completely customized with simple bash commands. If this is removed with the -client switch, then the whole thing loses its meaning.

**Reply by gokcehan** [reply]
> because it takes up unnecessary extra space on the screen

What do you mean by this? How exactly do you run the server? You are not manually run `lf -server` in a separate terminal, are you? Just to be sure, you are only supposed to run `lf` and the server is automatically started in the background. It should not occupy any space in the screen.

> it also runs more than I need in the process

Why is it an issue to have an extra process running? Note, when you define such custom commands instead of using the builtin copy-paste commands, you create a new process each time these custom commands are called. Is it not also an issue for you then to have these extra processes running?

**Comment by og900aero** [comment]
> > because it takes up unnecessary extra space on the screen
> 
> What do you mean by this? How exactly do you run the server? You are not manually run `lf -server` in a separate terminal, are you? Just to be sure, you are only supposed to run `lf` and the server is automatically started in the background. It should not occupy any space in the screen.

Run lfrun (because ueberzug). Then I get one windowd. When I want to copy while also seeing the original route, I have to open another window. In tiling windows manager, then two windows will appear, filling the entire screen, and the lf size will be reduced by half, since the same screen must now be shared by two windows. In my personal opinion, the server-client solution is not good for this reason either, so I prefer the tab solution. However, I love lf because I can customize it endlessly with bash language, so I solved it differently so that I don't need to open more windows, nor the tab function.

> > it also runs more than I need in the process
> 
> Why is it an issue to have an extra process running? Note, when you define such custom commands instead of using the builtin copy-paste commands, you create a new process each time these custom commands are called. Is it not also an issue for you then to have these extra processes running?

When I run lfrun, then run two prcess on the background (view btop):

1. lf -server
2. lf -command cd /home/Data

when i quit lf, the lf -server also runnning.
My custom copy-move command uses builtin copy-move command with little plus. It is not sognificant resources than lf -server and plus lf -command blabla process.

So would be great if lf- single mode also get customize it endlessly, than lf -server mode, because some people use it in one window, but still want to customize it properly.

**Reply by DusanLesan** [reply]
>  When I want to copy while also seeing the original route, I have to open another window. In tiling windows manager, then two windows will appear, filling the entire screen

This part is unclear to me because it sounds like you are intentionally opening second `lf` window even though you said `The reason I would use it is that I don't want to run multiple instances. 1 is enough`

> In my personal opinion, the server-client solution is not good for this reason either, so I prefer the tab solution

Could you describe what makes server-client solution unsuitable for your use case. Also what is a tab solution

> when i quit lf, the lf -server also runnning. 

You can set server to quit when no clients are left open with `autoquit`

> My custom copy-move command uses builtin copy-move command with little plus. It is not sognificant resources than lf -server and plus lf -command blabla process

So you are saying server takes up too much resources? I have never looked too deep in its resource usage apart from seeing that it uses less than a single terminal instance, with all GUI apps open I do not see it in a list of apps that use most memory and that its cpu usage does not even register in htop

**Reply by joelim-work** [reply]
Apologies for hijacking this discussion but it has been suggested in https://github.com/gokcehan/lf/issues/1761#issuecomment-2188046128 and https://github.com/gokcehan/lf/issues/912#issuecomment-1221303371 to make `autoquit` enabled by default. Does this make sense?

The reason is that some users don't appreciate leaving a server process behind after quitting `lf`, and the only downside I can think of is that the server may have to be started each time, but I normally don't notice any slowness caused by it.

**Comment by og900aero** [comment]
> This part is unclear to me because it sounds like you are intentionally opening second `lf` window even though you said `The reason I would use it is that I don't want to run multiple instances. 1 is enough`

Yes, I opened another lf, because in order to see the current directory where I'm going to copy from, I need another lf, since lf can't show two directories, like, midnight commander. This would be solved, for example, if, instead of the preview window, another window could be temporarily seen directory too, where you can also navigate in libraries, and then you could copy back and forth, like in midnight commander. Or the other solution is to use tabs. Examples include yazi, ranger, nnn, joshuto.

> Could you describe what makes server-client solution unsuitable for your use case. Also what is a tab solution
> 
Unsuitable Im wrote. In order to copy anything in such a way that I want to see both the source and target directory at the same time, I have to open another lf. Whereas with those file managers that manage tabs, I just switch to another tab and navigate to another place, so I can immediately browse both the source and destination folders.

> You can set server to quit when no clients are left open with `autoquit`

Thx. I just set this up now.

> So you are saying server takes up too much resources? I have never looked too deep in its resource usage apart from seeing that it uses less than a single terminal instance, with all GUI apps open I do not see it in a list of apps that use most memory and that its cpu usage does not even register in htop

No. Lf uses minimal resources, but I am a maximalist: "why should any program use even 1 process more or more memory, if I don't need it" I believe in the principle.

**Reply by DusanLesan** [reply]
How would having tabs help you considering you want to see the current directory where you're going to copy from? If you open another tab, you would leave source directory view in previous and you would not see it

As for having multiple panes like midnight commander, you are already having that by opening two `lf` clients side by side. It takes no more screen space than if it would handled by the app.

Regarding `why should any program use even 1 process more or more memory, if I don't need it` I 100% agree and I hate that less and less apps are making optional stuff really optional and we still need to have dependencies for them installed. That is why I love LF: it keeps the things simple by not including stuff like native support for tabs, multi-panes, previews, archive browsing, weird copy flows ets since most of it can be done externally by wrapping lf, having helper tasks... But I much more prefer having multiple tasks doing things in parallel than having one task especially considering that 1 task can be more resource demanding than several.

I see all your responses to me here separated from the thread where previous message was posted, so if possible please avoid separating responses to keep the issue cleaner and easy to skip off-topic conversations like this. 

On topic: Not sure if my opinion matters at all since I am not contributor, but I think this issue is invalid and should be closed considering that functionality exists in default flow but user decided to stop using server that handles remote commands
